### PR TITLE
[MAT-5916] Version QDM Measure

### PR DESCRIPTION
**Pull request by jkotanchik-SB**
## MADiE PR

Jira Ticket: [MAT-5916](https://jira.cms.gov/browse/MAT-5916)
(Optional) Related Tickets:

### Summary
The original versioning flow (FHIR based) worked fine with the exception of the FHIR Bundle creation and save at the end would fail when presented with a QDM Measure.

Reorganized and refactored the flow to branch based on whether the Measure is an instace of FhirMeasure, and assumed to be a QDM Measure if not.
Refactor versioning flow:
  - Separate QDM and FHIR measure versioning 
  - Use common validation method
  - Split out common versioning logic into its own method

Introduced an enum to identify the different kinds of version validation results instead of returning a ResponseEntity from the Service.
Refactor version validation:   
  - move common authn and validation checks to their own method
  - return validation result enum
  - simply invalid test case check lambda

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [x] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [x] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [x] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/cms/gov/madie/measure/resources/MeasureVersionController.java [hunk 1]
```diff
@@ -32,13 +34,17 @@ public ResponseEntity<Measure> createVersion(
...
+  public ResponseEntity<Void> checkValidVersion(
       @PathVariable("id") String id,
       @RequestParam String versionType,
       Principal principal,
-      @RequestHeader("Authorization") String accessToken)
-      throws Exception {
-    return versionService.checkValidVersioning(id, versionType, principal.getName(), accessToken);
+      @RequestHeader("Authorization") String accessToken) {
+    var validationResult =
+        versionService.checkValidVersioning(id, versionType, principal.getName(), accessToken);
```
Comment (RohitKandimalla) [inline src/main/java/cms/gov/madie/measure/resources/MeasureVersionController.java:43]: Wondering why we can't verify if measure is eligible for Versioning in the front end ?

Comment (jkotanchik-SB) [reply src/main/java/cms/gov/madie/measure/resources/MeasureVersionController.java:43]: It probably could. Depends on our approach to APIs.

Comment (RohitKandimalla) [reply src/main/java/cms/gov/madie/measure/resources/MeasureVersionController.java:43]: Ofcourse yeah. No hard opinion here, but just wanted to ask because this is introducing some latency.

**Review by RohitKandimalla — APPROVED** [review]

File: src/main/java/cms/gov/madie/measure/resources/MeasureVersionController.java (hunk 1 above)
Comment (jkotanchik-SB) [reply src/main/java/cms/gov/madie/measure/resources/MeasureVersionController.java:43]: I'm always split on what to do as we haven't solidified if madie will support concurrent measure editing.

Maybe a "shallow" check on the UI  that verifies versioning options and any conditions that will require user permission to continue, and a "deep" check that pulls the latest measure data for validation.

Comment (RohitKandimalla) [reply src/main/java/cms/gov/madie/measure/resources/MeasureVersionController.java:43]: Yeah, I didn't think of this. We already support users to share measures, So it is a possible scenario, and a validation check up front using API does reduce the window for possible concurrency issues.

Never worked on it, but love to implement some sort of DB triggers and events to handle the object changes dynamically on the UI. I guess that is a discussion for another day.
